table/tables/cache.go

```diff
@@ -30,74 +35,158 @@ var _ table.CachedTable = &cachedTable{}
 
 type cachedTable struct {
 	TableCommon
-	kv.MemBuffer
-	mu sync.RWMutex
+	cacheData atomic.Value
+	mu        sync.RWMutex
+	handle    StateRemote
+}
+
+func leaseFromTS(ts uint64) uint64 {
+	// TODO make this configurable in the following PRs
+	const defaultLeaseDuration time.Duration = 3 * time.Second
+	physicalTime := oracle.GetTimeFromTS(ts)
+	lease := oracle.GoTimeToTS(physicalTime.Add(defaultLeaseDuration))
+	return lease
 }
 
 func (c *cachedTable) TryGetMemcache(ts uint64) (kv.MemBuffer, bool) {
 	c.mu.RLock()
 	defer c.mu.RUnlock()
-	if c.isReadFromCache(ts) {
-		return c.MemBuffer, true
+	tmp := c.cacheData.Load()
+	if tmp == nil {
+		return nil, false
+	}
+	data := tmp.(*table.CacheData)
+	if data.Lease > ts {
```

Comment:
In MVCC transaction, a transaction comes with a concept of version (ts in TiDB)

```
begin; do sleep(3600); select * from small_table;
```

This transaction should read the data of 1h ago, even some other transaction modify the data later, it doesn't break the snapshot isolation semantic.

Reply:
What I mean is that for an old ts, it can get a wrong result from cache. Suppose the start ts is `ts1` which is a hour ago and the table is locked with an read lock with lease `ts2` . The cache version is `ts3` . `ts1 < ts3 < ts2` , so it will read data from cache instead of tikv. However we cannot guarantee that there no updates between `ts1` and `ts3` . So the select will get a newer data than expected.

Reply:
@lcwangchao The correct implement should use atomic operation, instead of lock based. Because the lock here do not prevent the "ABA" problem as you mentioned. 
The `ts` and `membuffer` together formed a `cacheData`. And when the plan is done, the snapshot of that `cacheData` is collected into the plan, and never change in the future. You can see how it's done in another PR: 

https://github.com/pingcap/tidb/pull/29443/files#diff-dc8c4c0ab93631d336a3e92583a64a1837135319cef7337758cd13ab86992086R144-R147

Reply:
Maybe we are referring the  same problem... That is my test result for this pr:

```
create table tc (id int);
alter table tc cache;
begin;  --- session1
select * from tc; --- session1, empty output

-- create another session session2
insert into tc values(1); --- session2
select * from tc; --- session2, trigger read lock

--- back to session1, still in txn
select * from tc; --- session1, empty output expected, but got result 1 in current implement
```

Reply:
yeah. this will modify in next pr. see https://github.com/pingcap/tidb/pull/29443/files#diff-dc8c4c0ab93631d336a3e92583a64a1837135319cef7337758cd13ab86992086R144-R147